### PR TITLE
Preset search

### PR DESCRIPTION
**Pull request by michaelzangl**
... to allow reviews

### CODE REVIEW

File: src/org/openstreetmap/josm/gui/tagging/presets/TaggingPreset.java [hunk 1]
```diff
@@ -634,4 +642,118 @@ public String getToolbarString() {
...
+        private static final String AND = " ";
+        private static final String OPENING_PAR = "(";
+        private static final String CLOSING_PAR = ")";
+
+        /**
+         * types used to search for OSM objects.
+         */
+        private static final String WAY = "WAY";
+        private static final String NODE = "NODE";
+        private static final String RELATION = "RELATION";
```
Comment (michaelzangl) [inline src/org/openstreetmap/josm/gui/tagging/presets/TaggingPreset.java]: The search parser has the exact same strings for parsing - wouldn't it be best to use them?

File: src/org/openstreetmap/josm/gui/tagging/presets/TaggingPreset.java [hunk 2]
```diff
@@ -634,4 +642,118 @@ public String getToolbarString() {
...
+         * For example, (type:WAY | type:NODE | type: RELATION) is returned by this method for the
+         * preset containing types : {@link TaggingPresetType#WAY}, {@link TaggingPresetType#NODE} and
+         * {@link TaggingPresetType#RELATION}.
+         * It is worth noting, that {@link TaggingPresetType#CLOSEDWAY} and {@link TaggingPresetType#MULTIPOLYGON}
+         * are display types and do not play a role in searching. So, they are simply converted to one of the
+         * standard types, namely CLOSEDWAY -> WAY, MULTIPOLYGON -> RELATION.
+         * @param ts types of the preset to be converted to the search query.
+         * @return a closed string of types combined with '|' character.
+         */
+        private String buildTypeQuery(Collection<TaggingPresetType> ts) {
```
Comment (michaelzangl) [inline src/org/openstreetmap/josm/gui/tagging/presets/TaggingPreset.java]: Don't allow null here. Use an empty collection instead.

File: src/org/openstreetmap/josm/gui/tagging/presets/TaggingPreset.java [hunk 3]
```diff
@@ -634,4 +642,118 @@ public String getToolbarString() {
...
+            return ts == null
+                    ? ""
+                    : ts.stream()
+                        .map(t -> t == TaggingPresetType.NODE
```
Comment (michaelzangl) [inline src/org/openstreetmap/josm/gui/tagging/presets/TaggingPreset.java]: Convert this lambda to a private method

Comment (michaelzangl) [inline src/org/openstreetmap/josm/gui/tagging/presets/TaggingPreset.java]: ... and use switch/case, with a `AssertionError` as default case.

File: src/org/openstreetmap/josm/gui/tagging/presets/TaggingPreset.java [hunk 4]
```diff
@@ -634,4 +642,118 @@ public String getToolbarString() {
...
+                    .collect(Collectors.joining(OR));
+        }
+
+        private String buildKeyValueTemplate(String key, String val) {
+            return new StringBuilder("\"")
+                    .append(key)
+                    .append("\"")
+                    .append("=")
+                    .append("\"")
+                    .append(val)
```
Comment (michaelzangl) [inline src/org/openstreetmap/josm/gui/tagging/presets/TaggingPreset.java]: You need to string escape that value

**Comment by don-vip** [comment]
Merged last summer

**Pull request closed without merging**